src/compiler/crystal/interpreter/repl.cr

```diff
@@ -94,22 +115,82 @@ class Crystal::Repl
     interpret_and_exit_on_error(node)
   end
 
-  private def interpret(node : ASTNode)
+  private def interpret(node : ASTNode, check_structural_changes : Bool = true)
     @main_visitor = MainVisitor.new(from_main_visitor: @main_visitor)
 
-    node = @program.normalize(node)
-    node = @program.semantic(node, main_visitor: @main_visitor)
+    begin
+      node = @program.normalize(node)
+      node = @program.semantic(node, main_visitor: @main_visitor)
+    rescue ex : Crystal::Error
+      # Don't show the error right away: they might have happened
+      # because of structural changes that aren't compatible with
+      # the current code.
+    end
+
+    if check_structural_changes && has_structural_changes?(node)
+      return handle_structural_changes
+    end
+
+    if ex
+      # If an error happened with the last node we tried to interpreter,
+      # don't consider it for future replays.
+      @past_nodes.pop
+      raise ex
+    end
+
     @interpreter.interpret(node, @main_visitor.meta_vars)
   end
 
+  private def handle_structural_changes(show_output : Bool = true)
+    puts "Structural change detected. Replyaing session...", &.dark_gray if show_output
+
+    begin
+      value = replay
+      puts "Session replayed!", &.dark_gray if show_output
+      value
+    rescue ex : Crystal::Error
+      puts "An error happened while replyaing with the recent structural changes:", &.yellow
```

Comment:
```suggestion
      puts "An error happened while replaying with the recent structural changes:", &.yellow
```

Reply:
Wow! It's not an accident then, it's encoded in my brain 😮